processing/src/main/java/io/druid/query/groupby/epinephelinae/Grouper.java

```diff
@@ -73,21 +75,30 @@
    *
    * @return result that is ok if the row was aggregated, not ok if a resource limit was hit
    */
-  AggregateResult aggregate(KeyType key);
+  default AggregateResult aggregate(KeyType key)
+  {
+    Preconditions.checkNotNull(key, "key");
+    return aggregate(key, hashFunction().apply(key));
+  }
 
   /**
    * Reset the grouper to its initial state.
    */
   void reset();
 
+  default Function<KeyType, Integer> hashFunction()
```

Comment:
Changed to ```ToIntFunction```.

> consider designing this system so that array-based aggregation could avoid boxing of the key also.

Yeah, I also have been thinking about that, but still don't have a good idea without huge code duplication. Do you have any idea?